Sources/Datadog/Tracing/Integrations/TracingWithLoggingIntegration.swift

```diff
@@ -73,26 +80,23 @@ internal struct TracingWithLoggingIntegration {
             )
         }
 
-        let threadName = getCurrentThreadName()
-
-        core.v1.scope(for: LoggingFeature.self)?.eventWriteContext { context, writer in
-            let log = logBuilder.createLogEvent(
-                date: date,
-                level: level,
-                message: message,
-                error: extractedError,
-                attributes: .init(
-                    userAttributes: userAttributes,
-                    internalAttributes: internalAttributes
-                ),
-                tags: [],
-                context: context,
-                threadName: threadName
-            )
-
-            if let log = log {
-                writer.write(value: log)
-            }
-        }
+        core.send(
+            message: .custom(
+                key: "log",
+                attributes: [
+                    "date": date,
+                    "loggerName": configuration.loggerName,
+                    "service": configuration.service,
+                    "threadName": Thread.current.dd.name,
+                    "message": message,
+                    "level": level,
+                    "error": extractedError,
+                    "userAttributes": userAttributes,
+                    "internalAttributes": internalAttributes,
+                    "sendNetworkInfo": configuration.sendNetworkInfo
+                ]
+            ),
```

Comment:
Unlike in #974, the type we pass here through `message` is quite verbose and full of `String` keys written and read on both sides. I wonder if it wouldn't be practical from testing POV to introduce sort of DTO object to handle it in both ends, e.g.:
```swift
// in Tracing

struct TracingLog: ToFeatureMessageConvertible {
   let date: Date
   let loggerName: String
   let service: String
   // ...

   var toFeatureMessage: FeatureMessage { /* ... */ }
}

// in Logging:

struct TracingLog: FromFeatureMessageConvertible {
   let date: Date
   let loggerName: String
   let service: String
   // ...

   init(from featureMessage: FeatureMessage) throws { 
      guard ... else { // here, validation happens (e.g. check for particular keys existence)
         throw InternalError("The message received through bus doesn't match Logging contract")
      }
      /// ...
   }
}
```

This could give us the ability of easy integration testing and documenting message specs as separate types, all watched in tests. Somewhere in V2 integration tests we might then do:
```swift
@testable import DatadogTracing
@testable import DatadogLogging

func testWhenProducingTracingLogInTracing_thenItCanBeReadInLogging() {
   let messageProduced: DatadogTracing.TracingLog = .mockRandom()
   let messageRead = DatadogLogging.TracingLog(from: messageProduced.toFeatureMessage)

   XCTAssertEqual(messageProduced.date, messageRead.date)
   XCTAssertEqual(messageProduced.loggerName, messageRead.loggerName)
   // ...
}
```
We might highly leverage this kind of testing when message passing requires some data transformation. The throwing initializer in `FromFeatureMessageConvertible` would give us additional opportunity of using `DD.telemetry` and watching if contracts are met everywhere and all the time.

This is somehow addressed by the _incomplete_, _partial_, _complete_ and _rejected_ message tests we do in this PR, with a difference that current design is asymmetric (we test received message, without producing the real one).

Even if not this way but some other, I think the "produce" and "receive" message would strongly benefit from more testing flavours, consistent error throwing flow and telemetry coverage that could be maintained easily. Any thoughts @maxep?

Reply:
That's a good point. I like the idea of being able to serialise/deserialise models on the bus, it would allow having data models defined in features with a simpler type check. Tho, I think we could leverage `Codable` for that instead of introducing new `Convertible` conformances. It would involve creating specific Encoder/Decoder for the bus, Features would then create their `Codable` objects and send/receive it on the bus. It's something I've already explored in my latest [oss (draft) project](https://github.com/maxep/swift-dynamic-object/tree/develop), and I think we could use a simple version of that with a pair of `AnyEncoder/AnyDecoder`.

About testing, the message 'producer' part is tested in `DDSpanTests` and `TracerTests. But indeed, it is lacking flavours for message failure on the bus: The bus is not able to say if the message failed because it's lacking a receiver, or because the message is malformed. This should be addressed so we can have valuable telemetry 👍

I will add both on dedicated PRs.

Reply:
My main concern is not that much about convenience, but rather clarity and tests of the contract between Features communication over the bus. I think this can be achieved in many ways, but at the highest level each may leverage most from introducing a type (_"When F1 sends `M`, then `F2` receives `M`"_), sort of DDD-approach.

But I like the idea of bringing convenience as well and what you propose with custom `Coder` is really interesting 👌. As long as we don't apply unnecessary perf penalty (like JSON-serialization would do), it's a valid option IMO 👍. Alternatively, we could think of code-generation to generate mirrored types for each message.

Maybe this topic is sth we should revisit when all bus communication is established (I still need to have a look at other PRs 👀).